src/main/scala/uk/gov/hmrc/bobby/domain/BobbyValidator.scala

```
@@ -24,38 +24,100 @@ object BobbyValidator {
+            (r.dependency.name.equals(dep.name) || r.dependency.name.equals("*")) &&
+            r.range.includes(version)
+        }
+        .sorted
 
-    // First sort the rules according to the precedence we impose, and partition to warnings and violations
-    val (violations, warnings) = bobbyRules.sorted.partition(r => r.effectiveDate.isBefore(now) || r.effectiveDate.equals(now))
+      matchingRules
+        .map { rule =>
+          if (rule.exemptProjects.contains(projectName))
+            BobbyExemption(rule)
+          else if (rule.effectiveDate.isBefore(now) || rule.effectiveDate.isEqual(now))
+            BobbyViolation(rule)
+          else
+            BobbyWarning(rule)
+        }
+        .sorted
+        .headOption
+        .getOrElse(BobbyOk)
+  }
+}
 
-    def matching(rules: List[BobbyRule]): List[BobbyRule] = rules.filter { rule =>
-      (rule.dependency.organisation.equals(dep.organization) || rule.dependency.organisation.equals("*")) &&
-        (rule.dependency.name.equals(dep.name) || rule.dependency.name.equals("*"))
-    }.filter(_.range.includes(version))
+sealed trait BobbyValidationResult {
+  def allMessages: List[Message]
+  def violations: List[Message]
+  def warnings: List[Message]
+  def exemptions: List[Message]
+  def hasViolations: Boolean
+  def hasWarnings: Boolean
+  def hasExemptions: Boolean
+  def hasNoIssues: Boolean
+}
 
-    // Always consider violations before warnings
-    matching(violations).headOption.map(BobbyViolation)
-      .orElse(matching(warnings).headOption.map(BobbyWarning))
-      .getOrElse(BobbyOk)
+object BobbyValidationResult {
 
-  }
+  def apply(messages: List[Message]): BobbyValidationResult =
+    Impl(messages.sortBy(_.moduleName))
+
+  private final case class Impl(allMessages: List[Message]) extends BobbyValidationResult {
+
+    override lazy val violations: List[Message] =
+      byResultName
```

How about just using the (unsafe) `byResultName.apply` here? I think it'd be better to blow up loudly than hide the error with a default

---

it's valid not to have any violations